File: wallet/android/app/src/main/kotlin/ai/banya/linkcpp/wallet/NodeTelemetry.kt
package ai.banya.linkcpp.wallet

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import java.io.File

data class DeviceStats(
    val ramTotalMb: Long = 0,
    val ramUsedMb: Long = 0,
    val ramAvailMb: Long = 0,
    val cpuLoad: Float = 0f,      // 0..1, aggregate
    val batteryTempC: Float = 0f, // celsius (thermal proxy)
    val charging: Boolean = false,
)

/** Best-effort device telemetry available to an unprivileged app. */
object NodeTelemetry {
    private var lastIdle = 0L
    private var lastTotal = 0L

    fun read(ctx: Context): DeviceStats {
        val am = ctx.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val mi = ActivityManager.MemoryInfo()
        am.getMemoryInfo(mi)
        val total = mi.totalMem / (1024 * 1024)
        val avail = mi.availMem / (1024 * 1024)

        val bi = ctx.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val temp = (bi?.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0) ?: 0) / 10f
        val status = bi?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
        val charging = status == BatteryManager.BATTERY_STATUS_CHARGING ||
            status == BatteryManager.BATTERY_STATUS_FULL

        return DeviceStats(
            ramTotalMb = total, ramUsedMb = total - avail, ramAvailMb = avail,
            cpuLoad = cpuLoad(), batteryTempC = temp, charging = charging,
        )
    }

    private fun cpuLoad(): Float = try {
        val line = File("/proc/stat").bufferedReader().use { it.readLine() } // "cpu u n s idle iowait irq ..."
        val v = line.trim().split(Regex("\\s+")).drop(1).map { it.toLong() }
        val idle = v[3] + v.getOrElse(4) { 0 }
        val total = v.sum()
        val dIdle = idle - lastIdle
        val dTotal = total - lastTotal
        lastIdle = idle; lastTotal = total
        if (dTotal <= 0) 0f else ((dTotal - dIdle).toFloat() / dTotal).coerceIn(0f, 1f)
    } catch (e: Exception) {
        0f
    }
}

/** Estimated resource profile for a (backend, mode) choice, grounded in on-device benchmarks. */
data class NodeProfile(
    val computeUnit: String,
    val memImpact: Float,   // 0..1
    val thermal: Float,     // 0..1
    val performance: Float, // 0..1
    val tokPerSec: Int,     // decode tok/s (qwen2.5-0.5B Q8_0, measured/estimated)
    val note: String,
)

fun nodeProfile(backend: String, mode: String, s: Strings): NodeProfile {
    val local = mode == "local_shard"
    return when (backend) {
        "opencl" -> NodeProfile(
            s.t("profile.opencl.unit"), 0.55f, 0.60f,
            if (local) 0.95f else 0.42f, if (local) 72 else 28,
            s.t("profile.opencl.note"),
        )
        "vulkan" -> NodeProfile(
            s.t("profile.vulkan.unit"), 0.55f, 0.55f,
            if (local) 0.72f else 0.46f, if (local) 75 else 32,
            s.t("profile.vulkan.note"),
        )
        else -> NodeProfile(
            s.t("profile.cpu.unit"), 0.35f, 0.85f,
            if (local) 0.78f else 0.55f, if (local) 100 else 39,
            s.t("profile.cpu.note"),
        )
    }
}

File: wallet/ios/App/NodeTelemetry.swift
import Foundation
import UIKit
import Darwin

/// Best-effort device telemetry available to an unprivileged iOS app.
struct DeviceStats {
    var ramTotalMb: UInt64 = 0
    var ramUsedMb: UInt64 = 0
    var ramAvailMb: UInt64 = 0
    var cpuLoad: Float = 0        // 0..1 aggregate
    var thermal: Float = 0        // 0..1 (from ProcessInfo.thermalState)
    var thermalWord: String = "thermal.low"   // localization key, resolved in the view
    var charging: Bool = false
}

/// Reads aggregate CPU/RAM/thermal/battery on iOS via mach + ProcessInfo + UIDevice.
final class NodeTelemetry {
    static let shared = NodeTelemetry()
    private var lastBusy: UInt64 = 0
    private var lastTotal: UInt64 = 0

    func read() -> DeviceStats {
        var s = DeviceStats()
        s.ramTotalMb = ProcessInfo.processInfo.physicalMemory / (1024 * 1024)
        let (usedMb, availMb) = memory()
        s.ramUsedMb = usedMb
        s.ramAvailMb = availMb
        s.cpuLoad = cpuLoad()

        switch ProcessInfo.processInfo.thermalState {
        case .nominal:  s.thermal = 0.25; s.thermalWord = "thermal.low"
        case .fair:     s.thermal = 0.50; s.thermalWord = "thermal.medium"
        case .serious:  s.thermal = 0.80; s.thermalWord = "thermal.high"
        case .critical: s.thermal = 1.00; s.thermalWord = "thermal.critical"
        @unknown default: s.thermal = 0.25; s.thermalWord = "thermal.low"
        }

        UIDevice.current.isBatteryMonitoringEnabled = true
        let bs = UIDevice.current.batteryState
        s.charging = (bs == .charging || bs == .full)
        return s
    }

    /// Aggregate CPU busy fraction from mach host_cpu_load_info tick deltas.
    private func cpuLoad() -> Float {
        var info = host_cpu_load_info()
        var count = mach_msg_type_number_t(MemoryLayout<host_cpu_load_info>.stride / MemoryLayout<integer_t>.stride)
        let kr = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                host_statistics(mach_host_self(), HOST_CPU_LOAD_INFO, $0, &count)
            }
        }
        guard kr == KERN_SUCCESS else { return 0 }
        let user = UInt64(info.cpu_ticks.0)
        let system = UInt64(info.cpu_ticks.1)
        let idle = UInt64(info.cpu_ticks.2)
        let nice = UInt64(info.cpu_ticks.3)
        let busy = user + system + nice
        let total = busy + idle
        defer { lastBusy = busy; lastTotal = total }
        let dBusy = busy &- lastBusy
        let dTotal = total &- lastTotal
        guard lastTotal > 0, dTotal > 0 else { return 0 }
        return min(max(Float(dBusy) / Float(dTotal), 0), 1)
    }

    /// (usedMb, availMb) from mach VM statistics; used = active + wired + compressed.
    private func memory() -> (UInt64, UInt64) {
        var stats = vm_statistics64()
        var count = mach_msg_type_number_t(MemoryLayout<vm_statistics64>.stride / MemoryLayout<integer_t>.stride)
        let kr = withUnsafeMutablePointer(to: &stats) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                host_statistics64(mach_host_self(), HOST_VM_INFO64, $0, &count)
            }
        }
        guard kr == KERN_SUCCESS else { return (0, 0) }
        let page = UInt64(vm_kernel_page_size)
        let used = (UInt64(stats.active_count) + UInt64(stats.wire_count)
            + UInt64(stats.compressor_page_count)) * page
        let free = (UInt64(stats.free_count) + UInt64(stats.inactive_count)) * page
        return (used / (1024 * 1024), free / (1024 * 1024))
    }
}

/// Estimated resource profile for a (backend, mode) choice on Apple Silicon.
/// iOS GPU path is **MLX** (Metal) — not OpenCL/Vulkan (which are the Android/Adreno path).
struct NodeProfile {
    let computeUnit: String
    let memImpact: Float   // 0..1
    let thermal: Float     // 0..1
    let performance: Float // 0..1
    let tokPerSec: Int     // decode tok/s (qwen2.5-0.5B Q8_0, estimated on Apple Silicon)
    let note: String
    let estimated: Bool    // iOS numbers are estimates, not on-device measured
}

@MainActor
func nodeProfile(backend: String, mode: String) -> NodeProfile {
    let local = mode == "local_shard"
    switch backend {
    case "mlx":
        return NodeProfile(
            computeUnit: "GPU · Apple Silicon (MLX/Metal)",
            memImpact: 0.50, thermal: 0.55,
            performance: local ? 0.92 : 0.45, tokPerSec: local ? 110 : 40,
            note: Localizer.shared.t("nodeProfile.mlxNote"),
            estimated: true)
    default: // cpu
        return NodeProfile(
            computeUnit: "CPU · Apple Silicon (Accelerate)",
            memImpact: 0.35, thermal: 0.70,
            performance: local ? 0.75 : 0.40, tokPerSec: local ? 85 : 33,
            note: Localizer.shared.t("nodeProfile.cpuNote"),
            estimated: true)
    }
}

/// Backends selectable on iOS (GPU = MLX, plus CPU).
enum NodeBackend {
    static let all: [(id: String, label: String)] = [
        ("mlx", "GPU · MLX"),
        ("cpu", "CPU"),
    ]
    /// Accelerator category reported to the hub for reward/monitor grouping.
    static func accelerator(_ backend: String) -> String { backend == "cpu" ? "cpu" : "gpu" }
}
